File: reference/go/aep/delivery_store_test.go
package aep

import (
	"testing"
)

func TestInMemoryDeliveryStore_TrackAndAck(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	seq := store.Track("evt_001", "sub_01")
	if seq != 1 {
		t.Fatalf("expected sequence 1, got %d", seq)
	}

	if !store.IsPending("evt_001") {
		t.Fatal("expected event to be pending")
	}
	if store.IsAcknowledged("evt_001") {
		t.Fatal("expected event not to be acknowledged")
	}

	acked := store.Ack("evt_001")
	if !acked {
		t.Fatal("expected ack to succeed")
	}
	if !store.IsAcknowledged("evt_001") {
		t.Fatal("expected event to be acknowledged")
	}
	if store.IsPending("evt_001") {
		t.Fatal("expected event not to be pending after ack")
	}
}

func TestInMemoryDeliveryStore_NackIncrements(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	store.Track("evt_001", "sub_01")
	attempts, ok := store.Nack("evt_001")
	if !ok {
		t.Fatal("expected nack to succeed")
	}
	if attempts != 2 {
		t.Fatalf("expected attempts 2, got %d", attempts)
	}

	pending := store.GetPending()
	if len(pending) != 1 {
		t.Fatalf("expected 1 pending, got %d", len(pending))
	}
	if pending[0]["attempts"] != 2 {
		t.Fatalf("expected pending attempts 2, got %v", pending[0]["attempts"])
	}
}

func TestInMemoryDeliveryStore_DeadLetters(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	store.Track("evt_001", "sub_01")
	dlq := store.DeadLetter("evt_001", map[string]any{
		"error": map[string]any{
			"code":    "timeout",
			"message": "no ack",
		},
	})
	if dlq == nil {
		t.Fatal("expected deadLetter to return event")
	}

	payload, ok := dlq["payload"].(map[string]any)
	if !ok {
		t.Fatal("expected deadLetter payload")
	}
	if payload["attempts"] != 1 {
		t.Fatalf("expected attempts 1, got %v", payload["attempts"])
	}
	if payload["original_event_id"] != "evt_001" {
		t.Fatalf("expected original_event_id evt_001, got %v", payload["original_event_id"])
	}
	if store.IsPending("evt_001") {
		t.Fatal("expected event not to be pending after deadLetter")
	}
}

func TestInMemoryDeliveryStore_Stats(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	store.Track("evt_a", "sub_01")
	store.Track("evt_b", "sub_01")
	store.Ack("evt_a")
	store.Track("evt_c", "sub_02")
	store.DeadLetter("evt_c", map[string]any{})

	stats := store.GetStats()
	if stats["totalSequences"] != 3 {
		t.Fatalf("expected totalSequences 3, got %v", stats["totalSequences"])
	}
	if stats["pending"] != 1 {
		t.Fatalf("expected pending 1, got %v", stats["pending"])
	}
	if stats["acknowledged"] != 1 {
		t.Fatalf("expected acknowledged 1, got %v", stats["acknowledged"])
	}
	if stats["deadLettered"] != 1 {
		t.Fatalf("expected deadLettered 1, got %v", stats["deadLettered"])
	}
}

func TestInMemoryDeliveryStore_NackReturnsFalseForUnknown(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	_, ok := store.Nack("nonexistent")
	if ok {
		t.Fatal("expected nack to return false for unknown event")
	}
}

func TestInMemoryDeliveryStore_DeadLetterReturnsNilForUnknown(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	dlq := store.DeadLetter("nonexistent", nil)
	if dlq != nil {
		t.Fatal("expected deadLetter to return nil for unknown event")
	}
}

func TestInMemoryDeliveryStore_HasAttemptsRemaining(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	store.Track("evt_001", "sub_01")
	if !store.HasAttemptsRemaining("evt_001", 3) {
		t.Fatal("expected attempts remaining with max 3 and 1 attempt")
	}
	store.Nack("evt_001")
	store.Nack("evt_001")
	if store.HasAttemptsRemaining("evt_001", 3) {
		t.Fatal("expected no attempts remaining with max 3 and 3 attempts")
	}
}

func TestInMemoryDeliveryStore_GetPendingForSubscriptionFilters(t *testing.T) {
	store := NewInMemoryDeliveryStore(0, "stream_01")

	store.Track("evt_a", "sub_01")
	store.Track("evt_b", "sub_02")
	store.Track("evt_c", "sub_01")

	filtered := store.GetPendingForSubscription("sub_01")
	if len(filtered) != 2 {
		t.Fatalf("expected 2 pending for sub_01, got %d", len(filtered))
	}

	found := make(map[string]bool)
	for _, e := range filtered {
		found[e["eventId"].(string)] = true
	}
	if !found["evt_a"] || !found["evt_c"] {
		t.Fatalf("expected evt_a and evt_c in filtered results, got %v", found)
	}
	if found["evt_b"] {
		t.Fatal("expected evt_b not in filtered results")
	}
}

File: implementations/typescript/test/delivery-store.test.js
import assert from "node:assert/strict";
import test from "node:test";
import { InMemoryDeliveryStore } from "../src/delivery-store-memory.js";

test("InMemoryDeliveryStore tracks and acknowledges events", () => {
  const store = new InMemoryDeliveryStore();

  const seq = store.track("evt_001", "sub_01", { eventId: "evt_001", subscriptionId: "sub_01" });
  assert.equal(seq, 1);

  assert.equal(store.isPending("evt_001"), true);
  assert.equal(store.isAcknowledged("evt_001"), false);

  const acked = store.ack("evt_001");
  assert.equal(acked, true);
  assert.equal(store.isAcknowledged("evt_001"), true);
  assert.equal(store.isPending("evt_001"), false);
});

test("InMemoryDeliveryStore nacks and increments attempts", () => {
  const store = new InMemoryDeliveryStore();

  store.track("evt_001", "sub_01", {});
  const attempts = store.nack("evt_001");
  assert.equal(attempts, 2);

  const pending = store.getPending();
  assert.equal(pending.length, 1);
  assert.equal(pending[0].attempts, 2);
});

test("InMemoryDeliveryStore dead-letters exhausted events", () => {
  const store = new InMemoryDeliveryStore();

  store.track("evt_001", "sub_01", {});
  const dlq = store.deadLetter("evt_001", { error: { code: "timeout", message: "no ack" } });
  assert.ok(dlq);
  assert.equal(dlq.payload.attempts, 1);
  assert.equal(dlq.payload.original_event_id, "evt_001");
  assert.equal(store.isPending("evt_001"), false);
});

test("InMemoryDeliveryStore lists dead-lettered records", () => {
  const store = new InMemoryDeliveryStore();
  store.track("evt_001", "sub_01", {});
  store.deadLetter("evt_001", { error: { code: "timeout" } });

  const records = store.getDeadLettered();
  assert.equal(records.length, 1);
  assert.equal(records[0].eventId, "evt_001");
  assert.equal(records[0].subscriptionId, "sub_01");
  assert.equal(records[0].reason.error.code, "timeout");
});

test("InMemoryDeliveryStore provides stats", () => {
  const store = new InMemoryDeliveryStore();
  store.track("evt_a", "sub_01", {});
  store.track("evt_b", "sub_01", {});
  store.ack("evt_a");
  store.track("evt_c", "sub_02", {});
  store.deadLetter("evt_c", {});

  const stats = store.getStats();
  assert.equal(stats.totalSequences, 3);
  assert.equal(stats.pending, 1);
  assert.equal(stats.acknowledged, 1);
  assert.equal(stats.deadLettered, 1);
});

test("InMemoryDeliveryStore nack returns false for unknown events", () => {
  const store = new InMemoryDeliveryStore();
  assert.equal(store.nack("nonexistent"), false);
});

test("InMemoryDeliveryStore deadLetter returns null for unknown events", () => {
  const store = new InMemoryDeliveryStore();
  assert.equal(store.deadLetter("nonexistent"), null);
});

test("InMemoryDeliveryStore hasAttemptsRemaining checks max attempts", () => {
  const store = new InMemoryDeliveryStore();
  store.track("evt_001", "sub_01", {});
  assert.equal(store.hasAttemptsRemaining("evt_001", 3), true);
  store.nack("evt_001");
  store.nack("evt_001");
  assert.equal(store.hasAttemptsRemaining("evt_001", 3), false);
});

test("InMemoryDeliveryStore getPendingForSubscription filters correctly", () => {
  const store = new InMemoryDeliveryStore();
  store.track("evt_a", "sub_01", {});
  store.track("evt_b", "sub_02", {});
  store.track("evt_c", "sub_01", {});

  const filtered = store.getPendingForSubscription("sub_01");
  assert.equal(filtered.length, 2);
  assert.deepEqual(filtered.map((e) => e.eventId), ["evt_a", "evt_c"]);
});

test("InMemoryDeliveryStore persists subscriptions", () => {
  const store = new InMemoryDeliveryStore();
  const record = store.createSubscription({ id: "sub_1", filter: { types: "task.*" }, created_at: "2026-07-11T10:00:00Z" });
  assert.equal(record.id, "sub_1");
  assert.equal(store.getSubscription("sub_1").filter.types, "task.*");
  assert.equal(store.listSubscriptions().length, 1);
  assert.equal(store.deleteSubscription("sub_1"), true);
  assert.equal(store.getSubscription("sub_1"), null);
  assert.equal(store.deleteSubscription("sub_1"), false);
});
